File: typescript/packages/core/tests/harness-fixture-replay.test.ts
/**
 * Fixture-replay integration test for Harness (issue #3).
 *
 * Loads `fixtures/model_responses/harness/react_loop.jsonl` and drives a
 * `StandardHarness` with `LoopStrategy::ReAct`, asserting the loop:
 *   1. Dispatches the recorded tool call
 *   2. Loops to the next agent turn
 *   3. Returns `Success` with the recorded final response
 *
 * Must produce the same outcome as the Rust integration test.
 */

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";

import {
  AgentId,
  ModelAgent,
  ReplayModelInterface,
  SessionId,
  StandardHarness,
  newTask,
  type HarnessConfig,
  type ProviderInfo,
} from "../src/index.js";
import {
  AllowAllSandbox,
  AlwaysContinuePolicy,
  NoopContextManager,
  ScriptedToolRegistry,
} from "../src/harness/testing.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(__dirname, "..", "..", "..", "..");
const fixturePath = resolve(repoRoot, "fixtures/model_responses/harness/react_loop.jsonl");

const provider: ProviderInfo = {
  name: "anthropic",
  model_id: "fixture",
  context_window: 200_000,
};

describe("Harness fixture replay — react_loop.jsonl", () => {
  it("dispatches recorded tool call then completes consistently with Rust", async () => {
    const jsonl = readFileSync(fixturePath, "utf-8");
    const replay = ReplayModelInterface.fromJsonl(jsonl, provider);
    const agent = new ModelAgent(AgentId.of("fixture-agent"), replay);

    const toolRegistry = new ScriptedToolRegistry().push({
      kind: "success",
      content: "127.0.0.1 localhost",
    });

    const config: HarnessConfig = {
      agent,
      toolRegistry,
      sandbox: new AllowAllSandbox(),
      contextManager: new NoopContextManager(),
      terminationPolicy: new AlwaysContinuePolicy(),
      modelParams: { stop_sequences: [] },
    };
    const harness = new StandardHarness(config);

    const task = newTask("read /etc/hosts then summarize", SessionId.of("fixture-session"), {
      kind: "re_act",
      max_iterations: 5,
    });

    const result = await harness.run({ task });
    expect(result.kind).toBe("success");
    if (result.kind === "success") {
      expect(result.output).toBe("127.0.0.1 localhost");
      expect(result.turns).toBe(2);
      expect(result.usage.input_tokens).toBe(30); // 12 + 18
      expect(result.usage.output_tokens).toBe(14); // 8 + 6
    }
    expect(toolRegistry.callCount).toBe(1);
  });
});

File: go/spore-core/harness_fixture_replay_test.go
package sporecore

import (
	"context"
	"os"
	"path/filepath"
	"runtime"
	"testing"
)

func harnessFixturePath(t *testing.T) string {
	t.Helper()
	_, this, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	dir := filepath.Dir(this)
	return filepath.Join(dir, "..", "..", "fixtures", "model_responses", "harness", "react_loop.jsonl")
}

// Mirrors rust/crates/spore-core/tests/harness_fixture_replay.rs.
// Drives a StandardHarness backed by a ReplayModel-fed ModelAgent against
// the shared JSONL fixture, asserting the same outcome as Rust:
//
//	Success { output: "127.0.0.1 localhost", turns: 2, input_tokens: 30, output_tokens: 14 }
//	tool registry dispatched exactly once.
func TestHarnessReActLoopDispatchesToolThenCompletes(t *testing.T) {
	raw, err := os.ReadFile(harnessFixturePath(t))
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	replay, err := ParseReplayJSONL(string(raw), ProviderInfo{
		Name: "anthropic", ModelID: "fixture", ContextWindow: 200_000,
	})
	if err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	agent := NewModelAgent(AgentID("fixture-agent"), replay)

	reg := NewScriptedToolRegistry()
	reg.Push(ToolOutput{Kind: ToolOutputSuccess, Content: "127.0.0.1 localhost"})

	cfg := HarnessConfig{
		Agent:             agent,
		ToolRegistry:      reg,
		Sandbox:           AllowAllSandbox{},
		ContextManager:    NoopContextManager{},
		TerminationPolicy: AlwaysContinuePolicy{},
	}
	h := NewStandardHarness(cfg)

	task := NewTask(
		"read /etc/hosts then summarize",
		SessionID("fixture-session"),
		ReActStrategy(5),
	)

	r := h.Run(context.Background(), NewHarnessRunOptions(task))
	if r.Kind != RunSuccess {
		t.Fatalf("expected Success, got %+v", r)
	}
	if r.Output != "127.0.0.1 localhost" {
		t.Fatalf("output = %q", r.Output)
	}
	if r.Turns != 2 {
		t.Fatalf("turns = %d, want 2", r.Turns)
	}
	if r.Usage.InputTokens != 30 {
		t.Fatalf("input tokens = %d, want 30", r.Usage.InputTokens)
	}
	if r.Usage.OutputTokens != 14 {
		t.Fatalf("output tokens = %d, want 14", r.Usage.OutputTokens)
	}
	if reg.CallCount.Load() != 1 {
		t.Fatalf("tool registry call count = %d, want 1", reg.CallCount.Load())
	}
}
